rclcpp/include/rclcpp/executor.hpp

```diff
@@ -156,9 +156,11 @@ class Executor
   {
     std::shared_ptr<void> message = subscription->create_message();
     bool taken = false;
-    rmw_take(subscription->subscription_handle_, message.get(), &taken);
-    if (taken) {
-      subscription->handle_message(message);
+    rmw_ret_t status = rmw_take(subscription->subscription_handle_, message.get(), &taken);
+    if (status == RMW_RET_OK) {
+      if (taken) {
+        subscription->handle_message(message);
+      }
     } else {
```

Comment:
The patch should be correct. If the function returns successful (which means without an error) it might still not have taken a message. Then neither the handle function need to be invoked nor an exception thrown.